processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -821,6 +829,14 @@ public void close() throws IOException
                                       ? null
                                       : Ints.asList(dims[i]);
           forwardDimWriters.get(i).write(listToWrite);
+          if (listToWrite == null || listToWrite.isEmpty()) {
+            // empty row; add to the nullRows bitmap
+            nullRowsList.get(i).add(rowCount);
+          } else if (dimHasNullFlags.get(i) && listToWrite.size() == 1 && listToWrite.get(0) == 0) {
+            // If this dimension has the null/empty str in its dictionary, a row with a single-valued dimension
+            // that matches the null/empty str's dictionary ID should also be added to nullRowsList.
+            nullRowsList.get(i).add(rowCount);
+          }
```

Comment:
Looked into this a bit, I think this "null bitmap" code in IndexMerger/IndexMergerV9 is handling two things:

- IncrementalIndexAdapter doesn't build bitmap indexes for null fully, it doesn't take into account empty or missing rows:
https://github.com/jon-wei/druid/commit/59e109021e283a1b6784a55a82b3fa2600e190c2

You can see that if you disable all of the "null bitmap" logic and run IndexMergerTest.testPersistWithDifferentDims()

- The "null bitmap" code also handles cases where two segments are merged and a dimension was only present in one, you can see that behavior by disable the "null bitmap" code and running IndexMergerTest.testDisjointDimMerge() 

If I disable the code pointed to by this comment thread in IndexMerger, I think some of the recently introduced filter tests on multival dimensions fails; enabling the "null bitmap" adjustment code in IndexIO.convertV8toV9() for multivalued dimensions as well as single valued dimensions seems to have the same effect as this change in question (but I haven't tested this extensively, only ran SelectorFilterTest)